python/ray/_private/runtime_env/mpi.py

```diff
@@ -0,0 +1,76 @@
+import sys
+import logging
+import argparse
+from typing import List, Optional
+from ray._private.runtime_env.context import RuntimeEnvContext
+from ray._private.runtime_env.plugin import RuntimeEnvPlugin
+import subprocess
+
+default_logger = logging.getLogger(__name__)
+
+
+class MPIPlugin(RuntimeEnvPlugin):
+    priority = 90
+    name = "mpi"
+
+    def modify_context(
+        self,
+        uris: List[str],  # noqa: ARG002
+        runtime_env: "RuntimeEnv",  # noqa: F821 ARG002
+        context: RuntimeEnvContext,
+        logger: Optional[logging.Logger] = default_logger,  # noqa: ARG002
+    ) -> None:
+        mpi_config = runtime_env.mpi()
+        if mpi_config is None:
+            return
+        try:
+            proc = subprocess.run(
+                ["mpirun", "--version"], capture_output=True, check=True
+            )
+        except subprocess.CalledProcessError:
+            logger.error(
+                "Failed to run mpi run. Please make sure mpi has been installed"
+            )
+            raise
+
+        logger.info(f"Running MPI plugin\n {proc.stdout.decode()}")
+
+        from pathlib import Path
+
+        # mpirun -n 10 python mpi.py worker_entry_func
+        worker_entry = mpi_config["worker_entry"]
+        assert Path(worker_entry).is_file()
+        cmds = (
+            ["mpirun"]
+            + mpi_config.get("args", [])
+            + [
+                context.py_executable,
+                str(Path(__file__).absolute()),
+                str(Path(worker_entry).absolute()),
+            ]
+        )
+        # Construct the start cmd
+        context.py_executable = " ".join(cmds)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Setup MPI worker")
```

Comment:
is this intentional?

Reply:
Yes.

Reply:
Do you mean the main function? or the parser?

This will will be used as the mpi entry point.

Reply:
main function. I don't see any main function from other plugins though. Maybe it should be a part of mpi_worker.py not here? (this means the function is executed when you import it?)

Reply:
MPIRUN is like a fork and the rest plugin doesn't have this.
The function won't execute when import since it checks main. If you import, `__name__` won't be `__main__`.


This piece of code is part of the plugin, that's why I put it here and it's simple.

But open to move if you insist.

Reply:
Hmm I see. I think this is a bit confusing when I read code first time. I prefer to move it to a separate file (something like mpi_start.py), but it is also okay if you add comments in details in the main block. E.g., "the plugin starts a subprocess that runs this main method. It is not executed as a part of normal plugin" or sth like that

Reply:
I thought about moving it to `python/ray/_private/workers/mpi_workers.py` but feel it just moves the code to far away from the place where it's used. I think split it and move it to the other file is better.